File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/CellLevelMultiselectDemo/CellLevelMultiselectDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "CellLevelMultiselectDemoMain.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxButtons"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxGridCardView"
#pragma link "cxGridDBCardView"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxMaskEdit"
#pragma link "cxSpinEdit"
#pragma link "cxTextEdit"
#pragma link "cxLookAndFeels"
#pragma link "cxDataStorage"
#pragma resource "*.dfm"
TCellLevelMultiselectDemoMainForm *CellLevelMultiselectDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TCellLevelMultiselectDemoMainForm::TCellLevelMultiselectDemoMainForm(TComponent* Owner)
  : TfmBaseForm(Owner)
{
  CreateSpreadSheet();
  TableView->Controller->SelectCells(
    TableView->Columns[1], TableView->Columns[5], 2, 12);
}
//---------------------------------------------------------------------------

int TCellLevelMultiselectDemoMainForm::SelectedRowCount()
{
  return TableView->Controller->SelectedRowCount;
}

int TCellLevelMultiselectDemoMainForm::SelectedColumnCount()
{
  return TableView->Controller->SelectedColumnCount;
}

int TCellLevelMultiselectDemoMainForm::GetSummOfSelection()
{
  Variant val;
  int Result = 0;
  for (int I = 0; I < SelectedRowCount(); I++)
    for (int J = 0; J < SelectedColumnCount(); J++){
      val = TableView->DataController->GetValue(
        TableView->Controller->SelectedRows[I]->RecordIndex,
        TableView->Controller->SelectedColumns[J]->Index);
      if (!VarIsNull(val))
        Result += (int)val;
    }
  return Result;
}

void __fastcall TCellLevelMultiselectDemoMainForm::TableViewSelectionChanged(
      TcxCustomGridTableView *Sender)
{
  TableView->ViewInfo->HeaderViewInfo->Update();
  lblSelectedRows->Caption =
    FloatToStrF(SelectedRowCount(), ffNumber, 15, 0);
  lblSelectedColumns->Caption  =
    FloatToStrF(SelectedColumnCount(), ffNumber, 15, 0);
  lblSelectedCells->Caption  =
    FloatToStrF(SelectedRowCount() * SelectedColumnCount(), ffNumber, 15, 0);
  lblSelectedSummary->Caption = FloatToStrF(GetSummOfSelection(), ffNumber, 15, 0);
}
//---------------------------------------------------------------------------

void __fastcall TCellLevelMultiselectDemoMainForm::TableViewCustomDrawColumnHeader(
      TcxGridTableView *Sender, TcxCanvas *ACanvas,
      TcxGridColumnHeaderViewInfo *AViewInfo, bool &ADone)
{
  InternalDrawItem(ACanvas, AViewInfo->Bounds,
    TcxBorders()<<bLeft<<bRight<<bTop<<bBottom, AViewInfo->Text,
      AViewInfo->Column->Selected);
  ADone = true;
}
//---------------------------------------------------------------------------

void __fastcall TCellLevelMultiselectDemoMainForm::TableViewCustomDrawIndicatorCell(
      TcxGridTableView *Sender, TcxCanvas *ACanvas,
      TcxCustomGridIndicatorItemViewInfo *AViewInfo, bool &ADone)
{
  if (dynamic_cast<TcxGridIndicatorRowItemViewInfo*>(AViewInfo) == 0)
    return;
  TcxCustomGridRow *ARecord =
    ((TcxGridIndicatorRowItemViewInfo*)AViewInfo)->GridRecord;
  InternalDrawItem(ACanvas, AViewInfo->ContentBounds,
    TcxBorders()<<bLeft<<bRight<<bBottom, IntToStr(ARecord->Index + 1),
      ARecord->Selected);
  ADone = true;
}
//---------------------------------------------------------------------------

void __fastcall TCellLevelMultiselectDemoMainForm::TableViewMouseDown(
      TObject *Sender, TMouseButton Button, TShiftState Shift, int X,
      int Y)
{
  TcxCustomGridHitTest *AHitTest = TableView->GetHitTest(X, Y);
  if (dynamic_cast<TcxGridColumnHeaderHitTest*>(AHitTest) == 0)
    return;
  TcxGridColumn *AColumn = ((TcxGridColumnHeaderHitTest*)AHitTest)->Column;
  TableView->BeginUpdate();
  try{
    if (Shift.Contains(ssShift)){
      if (FAnchorLinkedAreaLastColumn == NULL)
        TableView->Controller->ClearSelection();
      else
        SetColumnsSelected(TableView->Controller->CellSelectionAnchor,
          FAnchorLinkedAreaLastColumn, false);
      SetColumnsSelected(TableView->Controller->CellSelectionAnchor, AColumn, true);
    }
    else{
      if (Shift.Contains(ssCtrl))
        SetColumnsSelected(AColumn, AColumn, true);
      else
        TableView->Controller->SelectColumns(AColumn, AColumn);
      AColumn->Focused = true;
      TableView->Controller->CellSelectionAnchor = AColumn;
    }
    TableView->DataController->SelectAll();
    FAnchorLinkedAreaLastColumn = AColumn;
  }
  __finally{
    TableView->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void TCellLevelMultiselectDemoMainForm::CreateSpreadSheet()
{
  TableView->BeginUpdate();
  try{
    CreateColumns();
    CreateRows();
  }
  __finally{
    TableView->EndUpdate();
  }
}
//---------------------------------------------------------------------------

void TCellLevelMultiselectDemoMainForm::CreateColumns()
{
  TcxGridColumn *AColumn;
  for (int I = 0; I < ColumnCount; I++){
    AColumn = TableView->CreateColumn();
    AColumn->Caption = GetColumnCaption(I);
    AColumn->HeaderAlignmentHorz = taCenter;
    AColumn->DataBinding->ValueType = "Integer";
  }
}
//---------------------------------------------------------------------------

void TCellLevelMultiselectDemoMainForm::CreateRows()
{
  TableView->DataController->RecordCount = RecordCount;
  randomize();
  for (int I = 0; I < RecordCount; I++)
    for (int J = 0; J < ColumnCount; J++)
      TableView->DataController->SetValue(I, J, random(100));
}
//---------------------------------------------------------------------------

String TCellLevelMultiselectDemoMainForm::GetColumnCaption(int Index)
{
  const int AlphabetSymbolCount = (int)'Z' - (int)'A' + 1;
  String Result;
  if (div(Index, AlphabetSymbolCount).quot > 0)
     Result = GetColumnCaption(div(Index, AlphabetSymbolCount).quot - 1);
  else
    Result = "";
  Result = Result + char((int)'A' + div(Index, AlphabetSymbolCount).rem);
  return Result;
}
//---------------------------------------------------------------------------

void TCellLevelMultiselectDemoMainForm::InternalDrawItem(TcxCanvas *ACanvas,
  TRect ABounds, TcxBorders ABorders, String AText, bool AIsSelected)
{
  const TColor SelectedItemBorderColor = TColor(0xC56A31);
  TRect ATextRect = ABounds;
  InflateRect(&ATextRect, -2, -1);
  TcxStyle *AStyle;
  TColor ABorderColor;
  if (AIsSelected){
	AStyle = styleSelected;
  }
  else{
	AStyle = styleNormal;
  }
  TcxViewParams AParams;
  AParams.Bitmap = NULL;
  AParams.Color = AStyle->Color;
  ACanvas->FillRect(ABounds, AParams, False);
  ACanvas->Font = AStyle->Font;
  ACanvas->Font->Color = AStyle->TextColor;
  ACanvas->DrawTexT(AText, ATextRect,
	cxAlignHCenter | cxAlignVCenter, True);
}
//---------------------------------------------------------------------------

void TCellLevelMultiselectDemoMainForm::SetColumnsSelected(
  TcxGridColumn *AFromColumn, TcxGridColumn *AToColumn, bool ASelected)
{
  int AFromColIndex = AFromColumn->VisibleIndex;
  int AToColIndex = AToColumn->VisibleIndex;
  int I;
  if (AFromColIndex > AToColIndex){
    I = AToColIndex;
    AToColIndex = AFromColIndex;
    AFromColIndex = I;
  }
  TableView->BeginUpdate();
  try{
    for (int I = AFromColIndex; I <= AToColIndex; I++)
      TableView->VisibleColumns[I]->Selected = ASelected;
  }
  __finally{
    TableView->EndUpdate();
  }
  TableViewSelectionChanged(NULL);
}
//---------------------------------------------------------------------------


void __fastcall TCellLevelMultiselectDemoMainForm::TableViewStylesGetHeaderStyle(
	TcxGridTableView *Sender, TcxGridColumn *AColumn, TcxStyle *&AStyle)
{
  if (AColumn)
  {
	  if (AColumn->Selected)
		AStyle = styleSelected;
	  else
		AStyle = styleNormal;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/CellLevelMultiselectDemo/CellLevelMultiselectDemoMain.pas
unit CellLevelMultiselectDemoMain;

{$I cxVer.inc}

interface

uses
  Windows, Forms, Messages, SysUtils, Classes, ActnList, ImgList, Controls, Menus,
  StdCtrls, cxButtons, cxCheckBox, cxContainer, cxEdit, cxTextEdit,
  cxMaskEdit, cxSpinEdit, ExtCtrls, cxGridLevel, cxGridCustomTableView,
  cxGridCardView, cxGridDBCardView, cxClasses, cxControls,
  cxGridCustomView, cxGrid, ComCtrls, cxStyles, cxCustomData, cxGraphics,
  cxFilter, cxData, DB, cxDBData, cxDataStorage, cxLookAndFeelPainters,
  cxLookAndFeels, cxHyperLinkEdit, cxImageComboBox, cxDBLookupComboBox,
  cxMemo, cxGridTableView, BaseForm, cxLabel;

type
  TCellLevelMultiselectDemoMainForm = class(TfmBaseForm)
    Panel1: TPanel;
    Label1: TcxLabel;
    Label2: TcxLabel;
    Label3: TcxLabel;
    Label4: TcxLabel;
    lblSelectedRows: TcxLabel;
    lblSelectedColumns: TcxLabel;
    lblSelectedCells: TcxLabel;
    lblSelectedSummary: TcxLabel;
    styleSelected: TcxStyle;
    styleNormal: TcxStyle;
    Grid: TcxGrid;
    TableView: TcxGridTableView;
    Level: TcxGridLevel;
    procedure TableViewSelectionChanged(Sender: TcxCustomGridTableView);
    procedure TableViewCustomDrawIndicatorCell(Sender: TcxGridTableView;
      ACanvas: TcxCanvas; AViewInfo: TcxCustomGridIndicatorItemViewInfo;
      var ADone: Boolean);
    procedure TableViewStylesGetHeaderStyle(Sender: TcxGridTableView;
      AColumn: TcxGridColumn; var AStyle: TcxStyle);
    procedure TableViewMouseDown(Sender: TObject; Button: TMouseButton;
      Shift: TShiftState; X, Y: Integer);
  private
    FAnchorLinkedAreaLastColumn: TcxGridColumn;
    procedure CreateSpreadSheet;
    procedure CreateColumns;
    procedure CreateRows;
    function GetColumnCaption(Index: Integer): string;
    procedure SetColumnsSelected(AFromColumn, AToColumn: TcxGridColumn;
      ASelected: Boolean);
  public
    constructor Create(AOwner: TComponent); override;
  end;

var
  CellLevelMultiselectDemoMainForm: TCellLevelMultiselectDemoMainForm;

implementation

{$R *.dfm}

uses
  Variants, Dialogs, AboutDemoForm, cxVariants;

const
  RecordCount = 500;
  ColumnCount = 256;
  
constructor TCellLevelMultiselectDemoMainForm.Create(AOwner: TComponent);
begin
  inherited Create(AOwner);
  CreateSpreadSheet;
  TableView.Controller.SelectCells(
    TableView.Columns[1], TableView.Columns[5], 2, 12);
end;

procedure TCellLevelMultiselectDemoMainForm.TableViewSelectionChanged(
  Sender: TcxCustomGridTableView);

  function SelectedRowCount: Integer;
  begin
    Result := TableView.Controller.SelectedRowCount;
  end;

  function SelectedColumnCount: Integer;
  begin
    Result := TableView.Controller.SelectedColumnCount;
  end;

  function GetSummOfSelection: Integer;
  var
    I, J: Integer;
    val: Variant;
  begin
    Result := 0;
    for I := 0 to SelectedRowCount - 1 do
      for J := 0 to SelectedColumnCount - 1 do
      begin
        val := TableView.DataController.GetValue(
          TableView.Controller.SelectedRows[I].RecordIndex,
          TableView.Controller.SelectedColumns[J].Index);
        if not VarIsNull(val) then
          Inc(Result,  Integer(val));
      end;
  end;

begin
  TableView.ViewInfo.HeaderViewInfo.Update;
  lblSelectedRows.Caption :=
    FloatToStrF(SelectedRowCount, ffNumber, 15, 0);
  lblSelectedColumns.Caption  :=
    FloatToStrF(SelectedColumnCount, ffNumber, 15, 0);
  lblSelectedCells.Caption  :=
    FloatToStrF(SelectedRowCount * SelectedColumnCount, ffNumber, 15, 0);
  lblSelectedSummary.Caption := FloatToStrF(GetSummOfSelection, ffNumber, 15, 0);
end;

procedure TCellLevelMultiselectDemoMainForm.TableViewCustomDrawIndicatorCell(
  Sender: TcxGridTableView; ACanvas: TcxCanvas;
  AViewInfo: TcxCustomGridIndicatorItemViewInfo; var ADone: Boolean);
var
  AIndicatorViewInfo: TcxGridIndicatorRowItemViewInfo;
  ATextRect: TRect;
  AStyle: TcxStyle;
begin
  if not (AViewInfo is TcxGridIndicatorRowItemViewInfo) then
    Exit;
  ATextRect := AViewInfo.ContentBounds;
  AIndicatorViewInfo := AViewInfo as TcxGridIndicatorRowItemViewInfo;
  InflateRect(ATextRect, -2, -1);
  if AIndicatorViewInfo.GridRecord.Selected then
    AStyle := styleSelected
  else
    AStyle := styleNormal;
  Sender.LookAndFeelPainter.DrawHeader(ACanvas, AViewInfo.ContentBounds,
    ATextRect, [], cxBordersAll, cxbsNormal, taCenter, vaCenter,
    False, False, IntToStr(AIndicatorViewInfo.GridRecord.Index + 1),
    AStyle.Font, AStyle.TextColor, AStyle.Color);
  ADone := True;
end;

procedure TCellLevelMultiselectDemoMainForm.TableViewStylesGetHeaderStyle(
  Sender: TcxGridTableView; AColumn: TcxGridColumn; var AStyle: TcxStyle);
begin
  if AColumn = nil then
    Exit;
  if AColumn.Selected then
    AStyle := styleSelected
  else
    AStyle := styleNormal;
end;

procedure TCellLevelMultiselectDemoMainForm.CreateSpreadSheet;
begin
  TableView.BeginUpdate;
  try
    CreateColumns;
    CreateRows;
  finally
    TableView.EndUpdate;
  end;
end;

procedure TCellLevelMultiselectDemoMainForm.CreateColumns;
var
  I: Integer;
  AColumn: TcxGridColumn;
begin
  for I := 0 to ColumnCount - 1 do
  begin
    AColumn := TableView.CreateColumn;
    AColumn.Caption := GetColumnCaption(I);
    AColumn.HeaderAlignmentHorz := taCenter;
    AColumn.DataBinding.ValueType := 'Integer';
  end;
end;

procedure TCellLevelMultiselectDemoMainForm.CreateRows;
var
  I, J: Integer;
begin
  TableView.DataController.RecordCount := RecordCount;
  Randomize;
  for I := 0 to RecordCount - 1 do
    for J := 0 to ColumnCount - 1 do
      TableView.DataController.SetValue(I, J, Random(100));
end;

function TCellLevelMultiselectDemoMainForm.GetColumnCaption(Index: Integer): string;
const
  AlphabetSymbolCount = Integer('Z') - Integer('A') + 1;
begin
  if Index div AlphabetSymbolCount > 0 then
     Result := GetColumnCaption(Index div AlphabetSymbolCount - 1)
  else Result := '';
  Result := Result + char(Integer('A') + Index mod AlphabetSymbolCount);
end;

procedure TCellLevelMultiselectDemoMainForm.SetColumnsSelected(
  AFromColumn, AToColumn: TcxGridColumn; ASelected: Boolean);
var
  I: Integer;
  AFromColIndex, AToColIndex: Integer;
begin
  AFromColIndex := AFromColumn.VisibleIndex;
  AToColIndex := AToColumn.VisibleIndex;
  if AFromColIndex > AToColIndex then
  begin
    I := AToColIndex;
    AToColIndex := AFromColIndex;
    AFromColIndex := I;
  end;
  TableView.BeginUpdate;
  try
    for I := AFromColIndex to AToColIndex do
      TableView.VisibleColumns[I].Selected := ASelected;
  finally
    TableView.EndUpdate;
  end;
  TableViewSelectionChanged(nil);
end;

procedure TCellLevelMultiselectDemoMainForm.TableViewMouseDown(
  Sender: TObject; Button: TMouseButton; Shift: TShiftState; X,
  Y: Integer);
var
  AHitTest: TcxCustomGridHitTest;
  AColumn: TcxGridColumn;
begin
  AHitTest := TableView.GetHitTest(X, Y);
  if not (AHitTest is TcxGridColumnHeaderHitTest) then
    Exit;
  AColumn := (AHitTest as TcxGridColumnHeaderHitTest).Column;
  TableView.BeginUpdate;
  try
    if ssShift in Shift then
    begin
      if FAnchorLinkedAreaLastColumn = nil then
        TableView.Controller.ClearSelection
      else
        SetColumnsSelected(TableView.Controller.CellSelectionAnchor,
          FAnchorLinkedAreaLastColumn, False);
      SetColumnsSelected(TableView.Controller.CellSelectionAnchor, AColumn, True);
    end
    else
    begin
      if ssCtrl in Shift then
        SetColumnsSelected(AColumn, AColumn, True)
      else
        TableView.Controller.SelectColumns(AColumn, AColumn);
      AColumn.Focused := True;
      TableView.Controller.CellSelectionAnchor := AColumn;
    end;
    TableView.DataController.SelectAll;
    FAnchorLinkedAreaLastColumn := AColumn;
  finally
    TableView.EndUpdate;
  end;
end;

end.
